File: Java/src/main/java/data/Parameters.java
package data;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import javax.crypto.Mac;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Parameters {
    private ArrayList<Job> setOfJobs;
    private ArrayList<Task> setOfTasks;
    private ArrayList<Machine> setOfMachines;
    public int finalTimePoint = 250;
    private double alphaCompletionTime = 1;
    private double alphaTardiness = 10;
    private double alphaRobust = 0.1;

    private int timeWindowLength = 1;

    public Parameters(){
        this.setOfJobs = new ArrayList<>();
        this.setOfTasks = new ArrayList<>();
        this.setOfMachines = new ArrayList<>();
        this.finalTimePoint = getRoundUpToClosestFactor(this.finalTimePoint);
    }
    public void readData(String jsonPath) throws Exception {
        FileReader reader = new FileReader(jsonPath);
        JsonElement jsonElement = JsonParser.parseReader(reader);
        JsonObject jsonObject = jsonElement.getAsJsonObject();

        JsonArray machinesArray = jsonObject.getAsJsonArray("machines");
        JsonArray jobsArray = jsonObject.getAsJsonArray("jobs");

        for (JsonElement machineElement : machinesArray) {
            JsonObject machineObject = machineElement.getAsJsonObject();
            Machine machine = new Machine();
            machine.setId(machineObject.get("id").getAsInt());

            this.setOfMachines.add(machine);
        }

        for (JsonElement jobElement : jobsArray) {
            JsonObject jobObject = jobElement.getAsJsonObject();
            Job job = new Job();

            job.setId(jobObject.get("id").getAsInt());
            job.setDeadline(getRoundDownToClosestFactor(jobObject.get("deadline").getAsInt()));
            double priority;
            switch (jobObject.get("priority").getAsString()){
                case "LOW":
                    priority = 1;
                    break;
                case "MEDIUM":
                    priority = 4;
                    break;
                case "HIGH":
                    priority = 16;
                    break;
                default:
                    throw new Exception("Invalid priority definition");
            }

            job.setPriority(priority);
            ArrayList<Task> taskList = new ArrayList<>();

            JsonArray tasksArray = jobObject.getAsJsonArray("tasks");
            for(JsonElement taskElement : tasksArray){
                JsonObject taskObject = taskElement.getAsJsonObject();

                Task task = new Task();
                task.setId(taskObject.get("id").getAsInt());
                task.setPriority(priority);
                task.setProcessingTime(taskObject.get("processing_time").getAsInt());
                task.setDiscretizedProcessingTime(getRoundUpToClosestFactor(task.getProcessingTime()));
                task.setJobWhichBelongs(job);
                task.setPrecedingTaskId(taskObject.get("preceding_task").getAsInt());
                task.setSucceedingTaskId(taskObject.get("succeeding_task").getAsInt());
                task.setOldScheduleTime(taskObject.get("schedule").getAsInt());

                int machineId = taskObject.get("assigned_machine").getAsInt();
                Machine machine = this.setOfMachines.stream().filter(m -> m.getId() == machineId).findAny().orElse(null);
                if(machine == null){
                    throw new Exception("A task is not assigned to any machine");
                }

                machine.getSetOfAssignedTasks().add(task);
                task.setAssignedMachine(machine);


                taskList.add(task);
                this.setOfTasks.add(task);
            }

            job.setTasks(taskList);
            this.setOfJobs.add(job);
        }

        findPrecedenceRelationTasks();
    }

    private void findPrecedenceRelationTasks(){
        for(Task i : this.setOfTasks){
            i.setPrecedingTask(this.setOfTasks.stream().filter(task -> task.getId() == i.getPrecedingTaskId()).findAny().orElse(null));
            i.setSucceedingTask(this.setOfTasks.stream().filter(task -> task.getId() == i.getSucceedingTaskId()).findAny().orElse(null));
        }
    }

    public List<Integer> getSetOfTardyTimes(Task i){
        List<Integer> setOfTardyPoints = new ArrayList<>();
        for(int t = (i.getJobWhichBelongs().getDeadline() - i.getDiscretizedProcessingTime()) > 0 ? (i.getJobWhichBelongs().getDeadline() - i.getDiscretizedProcessingTime() + 1) : 0;
            t <= this.finalTimePoint; t++){
            setOfTardyPoints.add(t);
        }
        return setOfTardyPoints;
    }

    public int getRoundUpToClosestFactor(double val){
        return (int) (Math.ceil(val / (double) this.timeWindowLength));
    }

    public int getRoundDownToClosestFactor(double val){
        return (int) (Math.floor(val / (double) this.timeWindowLength) );
    }

    public ArrayList<Job> getSetOfJobs() {
        return setOfJobs;
    }

    public ArrayList<Task> getSetOfTasks() {
        return setOfTasks;
    }

    public ArrayList<Machine> getSetOfMachines() {
        return setOfMachines;
    }

    public int getFinalTimePoint() {
        return finalTimePoint;
    }

    public double getAlphaCompletionTime() {
        return alphaCompletionTime;
    }

    public double getAlphaRobust() {
        return alphaRobust;
    }

    public double getAlphaTardiness() {
        return alphaTardiness;
    }
}

File: Python/heuristic/data/parameters.py
from . import Task, Job, Machine
import json

class Parameters:
    set_of_jobs: list[Job]
    set_of_tasks: list[Task]
    set_of_machines: list[Machine]
    alpha_completion_time: float
    alpha_tardiness: float
    alpha_robust: float
    low_priority: float
    medium_priority: float
    high_priority: float
    scenario: dict

    def __init__(self):
        self.set_of_jobs = []
        self.set_of_tasks = []
        self.set_of_machines = []

        self.alpha_completion_time = 1
        self.alpha_tardiness = 10
        self.alpha_robust = 0.1

        self.low_priority = 1
        self.medium_priority = 4
        self.high_priority = 16

    def get_priority(self, priority: str):
        if priority == 'LOW':
            return self.low_priority
        elif priority == 'MEDIUM':
            return self.medium_priority
        elif priority == 'HIGH':
            return self.high_priority

    def read_data(self, path=None, json_file=None):
        if path is not None:
            with open(path, 'r') as f:
                scenario = json.load(f)

        if json_file is not None:
            scenario = json_file

        self.scenario = scenario

        for machine in scenario['machines']:
            id = int(machine['id'])
            processing_time_constant = float(machine['processing_time_constant'])
            type = machine['task_type_undertakes']
            mac = Machine(id, type, processing_time_constant)
            self.set_of_machines.append(mac)

        for job in scenario['jobs']:
            id = int(job['id'])
            deadline = int(job['deadline'])
            priority = job['priority']
            string_priority = priority
            priority = self.get_priority(string_priority)
            self.set_of_jobs.append(Job(id, deadline, priority, string_priority))

            for task in job['tasks']:
                id = int(task['id'])
                processing_time = float(task['processing_time'])
                machines_can_undertake_ids = [int(machine_id) for machine_id in task['machines_can_undertake']]
                machines_can_undertake = []
                for mac_id in machines_can_undertake_ids:
                    for mac in self.set_of_machines:
                        if mac.id == mac_id:
                            machines_can_undertake.append(mac)
                            break

                preceding_task_id = int(task['preceding_task'])
                succeeding_task_id = int(task['succeeding_task'])
                old_scheduled_time = float(task['scheduled_start_time'])
                t = Task(id, processing_time, preceding_task_id, succeeding_task_id, self.set_of_jobs[-1], priority,
                         old_scheduled_time)
                t.machines_can_undertake = machines_can_undertake
                for mac in machines_can_undertake:
                    mac.set_of_assigned_tasks.append(t)
                    t.processing_times[mac] = processing_time * mac.processing_time_constant

                self.set_of_tasks.append(t)
                self.set_of_jobs[-1].tasks.append(t)

        self._find_precedence_relations()

    def _find_precedence_relations(self):
        for task in self.set_of_tasks:
            if task.preceding_task_id != -1:
                for t in self.set_of_tasks:
                    if t.id == task.preceding_task_id:
                        task.preceding_task = t
                        break
            if task.succeeding_task_id != -1:
                for t in self.set_of_tasks:
                    if t.id == task.succeeding_task_id:
                        task.succeeding_task = t
                        break